### PR TITLE
Add dependency to redirect unauthenticated users to login and add webpages to display error responses.

### PR DESCRIPTION
**Pull request by kuries**
Parent: #25 

1) Redirect unauthorized users to `/login` endpoint.
2) Add extra web pages to handle HTTP 404 errors and responses from external URL invites.

- I added a dependency in the backend which acts as middleware and checks if the request has a valid session ID. If it doesn't, then it returns a "401 unauthorized request" as a response.

- To handle this on the client-side I created an Axios interceptor which checks for a 401 error in all the incoming responses. If present, it'll redirect the user to the login page.

Demo pages for 404 and External URL response messages:

error-404

error-404.png

image

External Url invite message.

### CODE REVIEW

**Comment by kuries** [comment]
error-404

Error 404 page

**Comment by kuries** [comment]
Currently, I tried to redirect unauthorised users in two ways:

1) The simple and easiest way is to add a dependency in the backend which checks for session ID and returns a 302 redirect response with location `/login`.  But this approach reloads the entire React App and goes against "SPA" way of doing things.
https://github.com/MatrixDoctors/matrix-cerberus/blob/5b8db584238b6ff94fe1579caa45436f82725976/backend/app/api/deps.py#L10-L14

2) The other approach (a little complex) is to return a 401 Unauthorized response from the backend and add axios interceptors which checks for a 401 on every response and then redirect from the client-side with a custom router from react-router.

https://github.com/MatrixDoctors/matrix-cerberus/blob/78b531874063cc77c88df28f372964de7ef47ed2/backend/app/api/deps.py#L10-L12

**Comment by kuries** [comment]
To redirect users on the login page we can do it in two ways.
1) A simple and easier way is to replace the window URL directly by changing `window.location.href`. This leads to a full page reload which might be against "SPA" nature.

2) A little more complex way is to create a custom router and supply a history object to it. This way we can pass this newly created history object to the axios interceptor and directly manipulate the custom router without triggering a full page reload. But this adds some unnecessary code which might possibly break in the future (not sure).
Reference: https://stackoverflow.com/a/70012117

@Half-Shot @Cadair 
Can you suggest which might be a good option in the long run?

File: frontend/src/HelperFunctions/customAxios.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    return response;
+  },
+  function (er) {
+    if (axios.isAxiosError(er)) {
+      if (er.response) {
+        if (er.response.status == 401) {
+          history.replace("/login");
+        }
+      }
+    }
```
Comment (Half-Shot) [inline frontend/src/HelperFunctions/customAxios.js:17]: Saves you a bit of nesting. You can use `foo?.bar` to say `if foo is undefined, don't try to access bar`. Very handy for these kinds of logic.

```suggestion
      if (axios.isAxiosError(er) && er.response?.status == 401) {
          history.replace("/login");
      }
```

File: frontend/src/HelperFunctions/customAxios.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import history from "./customHistory";
+
+const axiosInstance = axios.create();
+
+axiosInstance.interceptors.response.use(
+  function (response) {
+    return response;
+  },
+  function (er) {
+    if (axios.isAxiosError(er)) {
```
Comment (Half-Shot) [inline frontend/src/HelperFunctions/customAxios.js:11]: Wonder if this is a good thing to explicitly log, so we know why we've been navigated away?

File: frontend/src/Pages/ExternalUrl.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                            onClick={() => navigate(path)}
+                            className="sm:w-full lg:w-auto my-2 border rounded md py-4 px-8 text-center bg-indigo-600 text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-700 focus:ring-opacity-50">
+                                Take me there!
+                            </button>
+                        </div>
+                    </div>
+                </div>
+            </div>
+            <div>
+                <img src="https://i.ibb.co/ck1SGFJ/Group.png" />
```
Comment (Half-Shot) [inline frontend/src/Pages/ExternalUrl.js:50]: I spot another cheeky remote asset :p

File: frontend/src/Pages/LoginSuccess.js [hunk 4]
```diff
@@ -22,8 +22,16 @@ export default function LoginSuccess() {
             });
             authenticateWithOpenId(response.data);
         }
-        fetchData();
-        navigate('/');
+        try {
+            if(loginToken == null){
+                throw new Error("Invalid login token.");
+            }
+            fetchData();
```
Comment (Half-Shot) [inline frontend/src/Pages/LoginSuccess.js:29]: Always await these.
```suggestion
            await fetchData();
```

**Review by Half-Shot — CHANGES_REQUESTED** [review]

File: frontend/src/HelperFunctions/customAxios.js (hunk 1 above)
Comment (kuries) [reply frontend/src/HelperFunctions/customAxios.js]: Done.

File: frontend/src/Pages/ExternalUrl.js (hunk 3 above)
Comment (kuries) [reply frontend/src/Pages/ExternalUrl.js]: Done :D

**Review by Half-Shot — APPROVED** [review]
LGTM

**Review by Cadair — APPROVED** [review]

**Pull request merged**